inventory/inventory.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package inventory
+
+import (
+	"time"
+
+	"github.com/mendersoftware/log"
+	"github.com/mendersoftware/mender/client"
+	"github.com/pkg/errors"
+)
+
+type Attribute struct {
+	Name  string      `json:"name"`
+	Value interface{} `json:"value"`
+}
+
+type Data []Attribute
+
+func (id *Data) replaceAttributes(attr []Attribute) error {
+	iMap := make(map[string]Attribute, len(*id))
+	for _, ia := range *id {
+		iMap[ia.Name] = ia
+	}
+
+	for _, ia := range attr {
+		iMap[ia.Name] = ia
+	}
+
+	cnt := 0
+	for _, v := range iMap {
+		if len(*id) > cnt {
+			(*id)[cnt] = v
+		} else {
+			*id = append(*id, v)
+		}
+		cnt++
+	}
+	return nil
+}
+
+type Inventory struct {
+	sendTicker  *time.Ticker
+	freq        time.Duration
+	scriptsPath string
+	submitter   client.InventorySubmitter
+
+	canSend bool
+}
+
+func New(path string, freq time.Duration, iSubmitter client.InventorySubmitter) *Inventory {
+	ticker := time.NewTicker(freq)
+	tickChan := ticker.C
+
+	i := Inventory{
+		sendTicker:  ticker,
+		freq:        freq,
+		scriptsPath: path,
+		submitter:   iSubmitter,
+	}
+
+	go func() {
+		for {
+			select {
+			case <-tickChan:
+				log.Debug("ticker expired; ready to send inventory data")
+				i.canSend = true
```

Comment:
Here we're getting into advanced territory: This is not safe to do in C in multiple threads (Go routines basically are threads), because you haven't declared `canSend` to be an atomic type. This is `sig_atomic_t` in C, which usually boils down to something with `volatile`. Golang doesn't have `volatile`, and I have no clue what guarantees Golang makes about int types and thread safety, but I'm almost certain it will not work. The reason is that the variable will likely be in the processor cache, anything else would be extremely inefficient. And the cache is not shared between threads, hence this updating will not work in all cases (depends entirely on exact Golang optimization, definitely not something to rely on).

Golang seems to recommend [not sharing variables between threads at all](https://blog.golang.org/share-memory-by-communicating), and instead use channels all the time, and I tend to agree. I'm not sure if you can use a channel like a flag, but there is probably some way you can check whether the channel stream is empty.

Another alternative, which is most similar to what you already have, is to use the [atomic.Value](https://golang.org/pkg/sync/atomic/#Value) type in Golang, which does exactly what you need. I still consider channels safer if you are able to use them though.

And lastly, the traditional alternative would be to use a [mutex](https://golang.org/pkg/sync/#Mutex) to protect this access, which should provide the synchronization and memory barriers you need to avoid the caching problem, but is a bit on the heavy side for just one flag.